beacon-chain/slasher/simulator/simulator.go

```diff
@@ -0,0 +1,274 @@
+package simulator
+
+import (
+	"context"
+	"time"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/async/event"
+	"github.com/prysmaticlabs/prysm/beacon-chain/blockchain"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/feed"
+	statefeed "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/beacon-chain/operations/slashings"
+	"github.com/prysmaticlabs/prysm/beacon-chain/slasher"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/stategen"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/crypto/bls"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/time/slots"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "simulator")
+
+// ServiceConfig for the simulator.
+type ServiceConfig struct {
+	Params                      *Parameters
+	Database                    db.SlasherDatabase
+	StateNotifier               statefeed.Notifier
+	AttestationStateFetcher     blockchain.AttestationStateFetcher
+	HeadStateFetcher            blockchain.HeadFetcher
+	StateGen                    stategen.StateManager
+	SlashingsPool               slashings.PoolManager
+	PrivateKeysByValidatorIndex map[types.ValidatorIndex]bls.SecretKey
+}
+
+// Parameters for a slasher simulator.
+type Parameters struct {
+	SecondsPerSlot         uint64
+	SlotsPerEpoch          types.Slot
+	AggregationPercent     float64
+	ProposerSlashingProbab float64
+	AttesterSlashingProbab float64
+	NumValidators          uint64
+	NumEpochs              uint64
+}
+
+// Simulator defines a struct which can launch a slasher simulation
+// at scale using configuration parameters.
+type Simulator struct {
+	ctx                   context.Context
+	slasher               *slasher.Service
+	srvConfig             *ServiceConfig
+	indexedAttsFeed       *event.Feed
+	beaconBlocksFeed      *event.Feed
+	sentAttSlashingFeed   *event.Feed
+	sentBlockSlashingFeed *event.Feed
+	sentProposerSlashings map[[32]byte]*ethpb.ProposerSlashing
+	sentAttesterSlashings map[[32]byte]*ethpb.AttesterSlashing
+	genesisTime           time.Time
+}
+
+// DefaultParams for launching a slasher simulator.
+func DefaultParams() *Parameters {
```

Comment:
These are defaults arbitrary for now but perhaps can customized by the caller. They are pretty decent defaults for testing out slasher with some high probability of slashings, high aggregation, and fast slot advancement